comit/src/expiries/config.rs

```diff
@@ -0,0 +1,289 @@
+//! This module provides functionality for calculating the duration of actions
+//! required to determine the transition period from one swap state to the next.
+
+use std::f64::consts;
+use time::{prelude::*, Duration};
+
+// TODO: From somewhere within the system we need to return to the
+// user a transaction fee to use for each of the transactions (deploy,
+// fund, refund, redeem). In this module we assume the fee is set to
+// the suggested amount. We rely on this assumption for the
+// confirmation time calculations in this module to be correct.
+//
+// For Ethereum:
+//
+// gas limit: blockchain-contracts already implements gas limit functionality
+//            for each Ethereum transaction type.
+// gas price: current gas price can be learned from the Ethereum blockchain,
+//            for example from geth or infura.
+// modifier:  We could use a pre-configured modifier, as is widely done in the
+//            Ethereum ecosystem for slow, medium, fast.
+//
+// For Bitcoin:
+//
+// With a configured 'include within N blocks' we can use bitcoind via
+// bitcoin-cli, by calling:
+//
+//  `bitcoin-cli estimatesmartfee N_BLOCKS`
+//
+// CAVEAT: The fee estimator relies on an active mempool with good uptime to
+// watch fee activity on the network, that in turn implies that our calculations
+// herein are only as good as the fee estimator of the Bitcoin connector.
+
+const BITCOIN_BLOCK_TIME_SECS: usize = 600; // 10 minutes, average Bitcoin block time.
+const ETHEREUM_BLOCK_TIME_SECS: usize = 20; // Conservative Ethereum block time.
+const BITCOIN_MINE_WITHIN_N_BLOCKS: usize = 3; // Value arbitrarily picked.
+const ETHEREUM_MINE_WITHIN_N_BLOCKS: usize = 3;
+
+// TODO: Add support for lightning.
+#[derive(Clone, Copy, Debug)]
+pub enum Protocol {
+    Herc20Hbit,
+    HbitHerc20,
+}
+
+impl Protocol {
+    pub fn config(&self) -> Config {
+        // Since we are targeting Nectar and the App we can set these.
+        let alice = ActorType::Human;
```

Comment:
It's an interesting concept that you came up with. However, I am not sure it's appropriate. As @thomaseizinger mentioned on the mailing list, we want to move the app to a full automation.
Which means (to my understanding) that once you take an order, you only need to sign once with your hardware wallet and the rest will be handled by the app (bot).

More importantly than knowing who is behind the screen (or not), I'd be keen to have the time to act fair and balanced in the context of the "option problem". Each party should be able to watch the market and decide to proceed or abort for the same period of time.

Interestingly, Alice has two periods to do so:
A1. Before funding (no funds locked)
A2. Before redeeming (A & B funds locked)

And Bob only has one:
B1. Before funding (A funds locked)

It'd be interesting to say for example:

```
A1+A2 = B1: 
```
.e.g. A1 = 30min, A2 = 30min and B1 = 60min.

However, in terms of advantages/penalty, you could think:
- A1 gives least advantages as no funds are locked
- A2 gives a bit of advantage has both parties have funds locked, Alice has the choice but Bob has his funds locked for the least amount of time
- B1 gives the most advantage as it is risk free for Bob but Alice would be penalised by having her funds locked

It's basically some game theory (and I know nothing about it) and the question of what is the more advantageous:
- having the upper hand
- having more time to decide
- having funds locked for shorter time

etc...

I am guessing there is no easy answer unless we have stats that shows how many time people abort and in which conditions.

Reply:
Interesting thoughts. I was under the impression that only A2 constitutes the option problem.

Reply:
> Interesting thoughts. I was under the impression that only A2 constitutes the option problem.

Hum, ok maybe it's not a real "option problem" but B1 definitely gives Bob some upper hand as he can decide not to proceed depending on market direction and Alice would end up with assets locked for a long time.

But then, of course if he does decide to proceed as the market is going in the right direction for him, Alice may decide not to redeem as the market may be going in the bad direction for her.

So more Bob waits before funding (assume he eventually funds), more he risks for Alice to actually not proceed as she would also be watching the market.
Alice can watch the market during Atot = A1 + B1 + A2.
If Bob makes B1 long, then yes he gives himself some advantage to see where the market go but he also increases Atot allowing Alice to have even more time to watch the market (and likely decide in the opposite direction than Bob).

I hope it makes sense :) it's hard to make it clear.

Reply:
That's clear, it is for the reason you outline (i.e., Bob waiting helps and hinders) that I do not think we need consider B1. IMO we should limit our concern to the true option problem and forget the rest, just assume honest actors who start the swap as soon as possible and act in a 'rational' manner. Re: human/bot, perhaps these should be renamed to app/nectar but I feel that it is still useful to have two values here, even if we decide to set them to the same value for now. This is the same reasoning behind having two time to mine values (one for fund and one for redeem). The idea is to make sure we have all the knobs we need to correctly model the expiries, not to make sure we have the values correct.

Reply:
Thanks for tagging me on this @D4nte !

Maybe the `ActorType` abstraction is just too broad? Given the idea about more automation on the mailing list, only one out of two actions is performed by a human, hence maybe this should be broken down to _per action_?

What we also may want to consider (in addition to automating redeeming) is the idea of "pre-signing" the fund transaction and let cnd broadcast it at the right point in time (assuming the fees are still useful). This would allow users to trade against each other (i.e. no nectar) without having to attend the UI all the time. In that case, all transactions would be broadcasted by a machine and hence time-to-act is not that relevant anymore.

For now, I'd be interested in how big of difference this actually makes?
How big is TTA compared to the whole swap duration? Because if it is rather small, we might as well just ignore it for now in the face of the idea that we will eventually automate most of it.